File: go/tutorials/tour-of-orchestration-go/events/service.go
package events

import (
	"fmt"
	"github.com/google/uuid"
	"github.com/restatedev/sdk-go"
)

type PaymentRequest struct {
	Amount     int    `json:"amount"`
	Currency   string `json:"currency"`
	CustomerId string `json:"customerId"`
	OrderId    string `json:"orderId"`
}

type PaymentResult struct {
	Success       bool   `json:"success"`
	TransactionId string `json:"transactionId,omitempty"`
	ErrorMessage  string `json:"errorMessage,omitempty"`
}

type ConfirmationRequest struct {
	Id     string        `json:"id"`
	Result PaymentResult `json:"result"`
}

// Mock function to initiate payment
func initPayment(req PaymentRequest, paymentId string) (string, error) {
	fmt.Printf(">>> Initiating external payment %s\n", paymentId)
	fmt.Printf("  Confirm the payment via:\n")
	fmt.Printf("  curl localhost:8080/Payments/Confirm --json '{\"id\": \"%s\", \"result\": {\"success\": true, \"transactionId\": \"txn-123\"}}'\n", paymentId)
	return fmt.Sprintf("payRef-%s", uuid.New().String()), nil
}

type Payments struct{}

func (Payments) Process(ctx restate.Context, req PaymentRequest) (PaymentResult, error) {
	// Create awakeable to wait for webhook payment confirmation
	confirmation := restate.Awakeable[PaymentResult](ctx)

	// Initiate payment with external provider (Stripe, PayPal, etc.)
	_, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		_, err := initPayment(req, confirmation.Id())
		return restate.Void{}, err
	}, restate.WithName("pay"))
	if err != nil {
		return PaymentResult{}, err
	}

	// Wait for external payment provider to call our webhook
	result, err := confirmation.Result()
	if err != nil {
		return PaymentResult{}, err
	}

	return result, nil
}

// Webhook handler called by external payment provider
func (Payments) Confirm(ctx restate.Context, confirmation ConfirmationRequest) error {
	// Resolve the awakeable to continue the payment flow
	restate.ResolveAwakeable(ctx, confirmation.Id, confirmation.Result)
	return nil
}

File: rust/tutorials/tour-of-orchestration-rust/src/events/service.rs
use restate_sdk::{Context, Service, Result, RestateError, Awakeable};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Deserialize)]
struct PaymentRequest {
    amount: i32,
    currency: String,
    customer_id: String,
    order_id: String,
}

#[derive(Serialize, Deserialize, Clone)]
struct PaymentResult {
    success: bool,
    transaction_id: Option<String>,
    error_message: Option<String>,
}

#[derive(Deserialize)]
struct ConfirmationRequest {
    id: String,
    result: PaymentResult,
}

// Mock function to initiate payment
async fn init_payment(req: &PaymentRequest, payment_id: &str) -> Result<String, RestateError> {
    println!(">>> Initiating external payment {}", payment_id);
    println!("  Confirm the payment via:");
    println!("  curl localhost:8080/Payments/confirm --json '{{\"id\": \"{}\", \"result\": {{\"success\": true, \"transactionId\": \"txn-123\"}}}}'", payment_id);
    Ok(format!("payRef-{}", Uuid::new_v4()))
}

#[derive(Service)]
struct Payments;

impl Payments {
    #[restate_sdk::handler]
    async fn process(&self, ctx: &mut Context, req: PaymentRequest) -> Result<PaymentResult, RestateError> {
        // Create awakeable to wait for webhook payment confirmation
        let confirmation: Awakeable<PaymentResult> = ctx.awakeable();

        // Initiate payment with external provider (Stripe, PayPal, etc.)
        let payment_id = confirmation.id().clone();
        ctx.run("pay", async move {
            init_payment(&req, &payment_id).await
        }).await?;

        // Wait for external payment provider to call our webhook
        confirmation.await
    }

    #[restate_sdk::handler]
    async fn confirm(&self, ctx: &mut Context, confirmation: ConfirmationRequest) -> Result<(), RestateError> {
        // Resolve the awakeable to continue the payment flow
        ctx.resolve_awakeable(&confirmation.id, confirmation.result)?;
        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let service = Payments;
    
    restate_sdk::serve(service)
        .bind("0.0.0.0:9080")
        .await?;
    
    Ok(())
}